### DISCUSSION TITLE
How to set STA_SSID1 when building custom images

### DISCUSSION THREAD

**Discussion by plu** [arendst/Tasmota#21632] [Ideas]
Hi,

I'd like to set `STA_SSID1` and `STA_PASS1` dynamically when building images. In the end I want to inject them either via environment variables, or by reading some file from `/run/secrets/secret`. I have tried various attempts, but without success. For example by setting it via `build_flags` in `platformio_tasmota_cenv.ini` (still using hardcoded values here for testing purposes):

```ini
[tasmota]
build_flags    = -DSTA_PASS1='"foo"'
                 -DSTA_SSID1='"bar"'
```

Then another attempt via the `PLATFORMIO_BUILD_FLAGS` environment variable, also adding parameters `-D...` etc.

But none of this was successful. I'd be super happy to read any ideas/pointers into the right direction. I just want to achieve setting these values without having them in any of the header files.

Thanks!

**Comment by sfromis** [comment]
See https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env.ini#L20-L21

**Comment by barbudor** [comment]
It won't work to  set STA_PASS1 and STA_SSID1 from the command lines options as they are defined in `my_user_config.`

The only supported way is to go through `user_config_override.h` with 
```
// -- Setup your own Wifi settings  ---------------
#undef  STA_SSID1
#define STA_SSID1         "YourSSID"             // [Ssid1] Wifi SSID

#undef  STA_PASS1
#define STA_PASS1         "YourWifiPassword"     // [Password1] Wifi password
```

Nothing prevent you to generate user_config_override.h with a script from environment variable, but you won't be able to do it from the command line unless you modify `my_user_config.h`

**Reply by barbudor** [reply]
Actually, you can propose a PR to change those lines:
https://github.com/arendst/Tasmota/blob/75c47b6948bc8028a9bd661a7eaafdc3d6675000/tasmota/my_user_config.h#L74-L75

to
```
ifndef STA_SSID1
#define STA_SSID1              ""                // [Ssid1] Wi-Fi SSID
#endif
#ifndef STA_PASS1
#define STA_PASS1              ""                // [Password1] Wi-Fi password
#endif
```
if accepted, that will allow to set those macros from the command line

**Reply by plu** [reply]
Thank you, this was the missing piece!

For anyone coming here and looking for the final solution:

`platformio_tasmota_cenv.ini`
```ini
[env:tasmota]
build_flags = ${env.build_flags}
              -DSTA_PASS1='"${sysenv.STA_PASS1}"'
              -DSTA_SSID1='"${sysenv.STA_SSID1}"'
```

```diff
diff --git a/tasmota/my_user_config.h b/tasmota/my_user_config.h
index 8bc62baae..167222ba4 100644
--- a/tasmota/my_user_config.h
+++ b/tasmota/my_user_config.h
@@ -71,8 +71,12 @@
 #define WIFI_DNS               "192.168.1.1"     // [IpAddress4] If not using DHCP set DNS1 IP address (might be equal to WIFI_GATEWAY)
 #define WIFI_DNS2              "0.0.0.0"         // [IpAddress5] If not using DHCP set DNS2 IP address (might be equal to WIFI_GATEWAY)
 
+#ifndef STA_SSID1
 #define STA_SSID1              ""                // [Ssid1] Wi-Fi SSID
+#endif
+#ifndef STA_PASS1
 #define STA_PASS1              ""                // [Password1] Wi-Fi password
+#endif
 #define STA_SSID2              ""                // [Ssid2] Optional alternate AP Wi-Fi SSID
 #define STA_PASS2              ""                // [Password2] Optional alternate AP Wi-Fi password
 #define WIFI_AP_PASSPHRASE     ""                // AccessPoint passphrase. For WPA2 min 8 char, for open use "" (max 63 char).
```

**Reply by sfromis** [reply]
If the build process should support using `build-flags` as a general mechanism for setting `#define` values, then *every* value in `my_user_config.h` would have to be wrapped in `#ifndef` / `#endif`. Unless this is done, or some other mechanism, the current state is pretty simple:

The build process **does not support** using `build-flags` the way you want, unless you make your own modifications. `STA_SSID` / `STA_PASS1` is only a tiny part of the broader issue.

**Comment by Jason2866** [comment]
Changed with merged PR #21640

**Reply by plu** [reply]
Thank you! :)

**Comment by Jason2866** [comment]
@plu Like this way!!
```python
[env:tasmota]
build_flags = ${env.build_flags}
              -DSTA_PASS1='"${sysenv.STA_PASS1}"'
              -DSTA_SSID1='"${sysenv.STA_SSID1}"'
```
